build: fix multiple definitions

arinc9 opened the pull request:
Fix GCC release 10 being unable to compile iproute-mptcp because of multiple definition of functions.

```
/usr/bin/ld: ipxfrm.o:(.bss+0x0): multiple definition of `filter'; ipmroute.o:(.bss+0x0): first defined here
/usr/bin/ld: xfrm_monitor.o:(.bss+0x0): multiple definition of `listen_all_nsid'; ipmonitor.o:(.bss+0x0): first defined here
/usr/bin/ld: libutil.a(utils.o):(.bss+0xc): multiple definition of `pretty'; tipc.o:tipc.c:28: first defined here
```

References: https://bugzilla.opensuse.org/1160244

Signed-off-by: Arınç ÜNAL <arinc.unal@gmail.com>

matttbe commented:
Thank you for this PR!

But I'm not sure to understand the errors. May you copy-paste the output from the compiler please?
Also the modifications are not related to MPTCP and I don't see these fixes upstream in https://git.kernel.org/pub/scm/network/iproute2/iproute2.git . Any idea why?

arinc9 commented:
Hey Matt. This issue does not exist on upstream. I've made this hacky fix to get iproute-mptcp to compile. I assume there are different attempts to mitigate this issue on upstream.

iproute-mptcp won't compile on `gcc (Ubuntu 10.3.0-1ubuntu1) 10.3.0`
```makefile
iproute-mptcp# ./configure && make
TC schedulers
 ATM    no

libc has setns: yes
SELinux support: no
ELF support: yes
libmnl support: yes
Berkeley DB: no
need for strlcpy: yes
libcap support: no

lib
    CC       libgenl.o
    CC       libnetlink.o
    AR       libnetlink.a
    CC       utils.o
    CC       rt_names.o
    CC       ll_map.o
    CC       ll_types.o
    CC       ll_proto.o
    CC       ll_addr.o
    CC       inet_proto.o
    CC       namespace.o
    CC       json_writer.o
    CC       json_print.o
    CC       names.o
    CC       color.o
    CC       bpf.o
bpf.c: In function ‘bpf_get_work_dir’:
bpf.c:779:49: warning: ‘__builtin___snprintf_chk’ output may be truncated before the last format character [-Wformat-truncation=]
  779 |  snprintf(bpf_wrk_dir, sizeof(bpf_wrk_dir), "%s/", mnt);
      |                                                 ^
In file included from /usr/include/stdio.h:866,
                 from bpf.c:14:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 2 and 4097 bytes into a destination of size 4096
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
    CC       exec.o
    CC       fs.o
    CC       dnet_ntop.o
    CC       dnet_pton.o
    CC       ipx_ntop.o
    CC       ipx_pton.o
    CC       mpls_ntop.o
    CC       mpls_pton.o
    AR       libutil.a

ip
    CC       ip.o
    CC       ipaddress.o
    CC       ipaddrlabel.o
    CC       iproute.o
    CC       iprule.o
    CC       ipnetns.o
    CC       rtm_map.o
    CC       iptunnel.o
    CC       ip6tunnel.o
    CC       tunnel.o
    CC       ipneigh.o
    CC       ipntable.o
    CC       iplink.o
    CC       ipmaddr.o
    CC       ipmonitor.o
    CC       ipmroute.o
    CC       ipprefix.o
    CC       iptuntap.o
    CC       iptoken.o
    CC       ipxfrm.o
    CC       xfrm_state.o
    CC       xfrm_policy.o
    CC       xfrm_monitor.o
    CC       iplink_dummy.o
    CC       iplink_ifb.o
    CC       iplink_nlmon.o
    CC       iplink_team.o
    CC       iplink_vcan.o
    CC       iplink_vxcan.o
    CC       iplink_vlan.o
    CC       link_veth.o
    CC       link_gre.o
    CC       iplink_can.o
    CC       iplink_xdp.o
    CC       iplink_macvlan.o
    CC       ipl2tp.o
    CC       link_vti.o
    CC       link_vti6.o
    CC       iplink_vxlan.o
    CC       tcp_metrics.o
    CC       iplink_ipoib.o
    CC       ipnetconf.o
    CC       link_ip6tnl.o
    CC       link_iptnl.o
    CC       link_gre6.o
    CC       iplink_bond.o
    CC       iplink_bond_slave.o
    CC       iplink_hsr.o
    CC       iplink_bridge.o
iplink_bridge.c: In function ‘bridge_print_opt’:
iplink_bridge.c:79:31: warning: ‘%s’ directive output may be truncated writing up to 31 bytes into a region of size 27
-Wformat-truncation=]
   79 |  snprintf(buf, len, "%.2x%.2x.%s", id->prio[0], id->prio[1], eaddr);
      |                               ^~                             ~~~~~
In file included from /usr/include/stdio.h:866,
                 from iplink_bridge.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 6 and 37 bytes into a destination of size 32
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
iplink_bridge.c:79:31: warning: ‘%s’ directive output may be truncated writing up to 31 bytes into a region of size 27
-Wformat-truncation=]
   79 |  snprintf(buf, len, "%.2x%.2x.%s", id->prio[0], id->prio[1], eaddr);
      |                               ^~                             ~~~~~
In file included from /usr/include/stdio.h:866,
                 from iplink_bridge.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 6 and 37 bytes into a destination of size 32
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
    CC       iplink_bridge_slave.o
    CC       ipfou.o
    CC       iplink_ipvlan.o
    CC       iplink_geneve.o
    CC       iplink_vrf.o
    CC       iproute_lwtunnel.o
    CC       ipmacsec.o
    CC       ipila.o
    CC       ipvrf.o
    CC       iplink_xstats.o
    CC       ipseg6.o
    CC       iplink_netdevsim.o
    CC       iplink_rmnet.o
    LINK     ip
/usr/bin/ld: ipxfrm.o:(.bss+0x0): multiple definition of `filter'; ipmroute.o:(.bss+0x0): first defined here
/usr/bin/ld: xfrm_monitor.o:(.bss+0x0): multiple definition of `listen_all_nsid'; ipmonitor.o:(.bss+0x0): first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:26: ip] Error 1
make: *** [Makefile:70: all] Error 2
```
With my fix iproute-mptcp compiles fine:
```makefile
arinc9-iproute-mptcp# ./configure && make
TC schedulers
 ATM    no

libc has setns: yes
SELinux support: no
ELF support: yes
libmnl support: yes
Berkeley DB: no
need for strlcpy: yes
libcap support: no

lib
    CC       libgenl.o
    CC       libnetlink.o
    AR       libnetlink.a
    CC       utils.o
    CC       rt_names.o
    CC       ll_map.o
    CC       ll_types.o
    CC       ll_proto.o
    CC       ll_addr.o
    CC       inet_proto.o
    CC       namespace.o
    CC       json_writer.o
    CC       json_print.o
    CC       names.o
    CC       color.o
    CC       bpf.o
bpf.c: In function ‘bpf_get_work_dir’:
bpf.c:779:49: warning: ‘__builtin___snprintf_chk’ output may be truncated before the last format character [-Wformat-truncation=]
  779 |  snprintf(bpf_wrk_dir, sizeof(bpf_wrk_dir), "%s/", mnt);
      |                                                 ^
In file included from /usr/include/stdio.h:866,
                 from bpf.c:14:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 2 and 4097 bytes into a destination of size 4096
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
    CC       exec.o
    CC       fs.o
    CC       dnet_ntop.o
    CC       dnet_pton.o
    CC       ipx_ntop.o
    CC       ipx_pton.o
    CC       mpls_ntop.o
    CC       mpls_pton.o
    AR       libutil.a

ip
    CC       ip.o
    CC       ipaddress.o
    CC       ipaddrlabel.o
    CC       iproute.o
    CC       iprule.o
    CC       ipnetns.o
    CC       rtm_map.o
    CC       iptunnel.o
    CC       ip6tunnel.o
    CC       tunnel.o
    CC       ipneigh.o
    CC       ipntable.o
    CC       iplink.o
    CC       ipmaddr.o
    CC       ipmonitor.o
    CC       ipmroute.o
    CC       ipprefix.o
    CC       iptuntap.o
    CC       iptoken.o
    CC       ipxfrm.o
    CC       xfrm_state.o
    CC       xfrm_policy.o
    CC       xfrm_monitor.o
    CC       iplink_dummy.o
    CC       iplink_ifb.o
    CC       iplink_nlmon.o
    CC       iplink_team.o
    CC       iplink_vcan.o
    CC       iplink_vxcan.o
    CC       iplink_vlan.o
    CC       link_veth.o
    CC       link_gre.o
    CC       iplink_can.o
    CC       iplink_xdp.o
    CC       iplink_macvlan.o
    CC       ipl2tp.o
    CC       link_vti.o
    CC       link_vti6.o
    CC       iplink_vxlan.o
    CC       tcp_metrics.o
    CC       iplink_ipoib.o
    CC       ipnetconf.o
    CC       link_ip6tnl.o
    CC       link_iptnl.o
    CC       link_gre6.o
    CC       iplink_bond.o
    CC       iplink_bond_slave.o
    CC       iplink_hsr.o
    CC       iplink_bridge.o
iplink_bridge.c: In function ‘bridge_print_opt’:
iplink_bridge.c:79:31: warning: ‘%s’ directive output may be truncated writing up to 31 bytes into a region of size 27 [-Wformat-truncation=]
   79 |  snprintf(buf, len, "%.2x%.2x.%s", id->prio[0], id->prio[1], eaddr);
      |                               ^~                             ~~~~~
In file included from /usr/include/stdio.h:866,
                 from iplink_bridge.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 6 and 37 bytes into a destination of size 32
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
iplink_bridge.c:79:31: warning: ‘%s’ directive output may be truncated writing up to 31 bytes into a region of size 27 [-Wformat-truncation=]
   79 |  snprintf(buf, len, "%.2x%.2x.%s", id->prio[0], id->prio[1], eaddr);
      |                               ^~                             ~~~~~
In file included from /usr/include/stdio.h:866,
                 from iplink_bridge.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:71:10: note: ‘__builtin___snprintf_chk’ output between 6 and 37 bytes into a destination of size 32
   71 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |        __glibc_objsize (__s), __fmt,
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |        __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~
    CC       iplink_bridge_slave.o
    CC       ipfou.o
    CC       iplink_ipvlan.o
    CC       iplink_geneve.o
    CC       iplink_vrf.o
    CC       iproute_lwtunnel.o
    CC       ipmacsec.o
    CC       ipila.o
    CC       ipvrf.o
    CC       iplink_xstats.o
    CC       ipseg6.o
    CC       iplink_netdevsim.o
    CC       iplink_rmnet.o
    LINK     ip
    CC       rtmon.o
    LINK     rtmon

tc
    CC       tc.o
    CC       tc_qdisc.o
    CC       tc_class.o
    CC       tc_filter.o
    CC       tc_util.o
    CC       tc_monitor.o
    CC       tc_exec.o
    CC       m_police.o
    CC       m_estimator.o
    CC       m_action.o
    CC       m_ematch.o
    YACC     emp_ematch.yacc.c
emp_ematch.y:11.1-14: warning: deprecated directive: ‘%error-verbose’, use ‘%define parse.error verbose’ [-Wdeprecated]
   11 | %error-verbose
      | ^~~~~~~~~~~~~~
      | %define parse.error verbose
emp_ematch.y:12.1-22: warning: deprecated directive: ‘%name-prefix "ematch_"’, use ‘%define api.prefix {ematch_}’ [-Wdeprecated]
   12 | %name-prefix "ematch_"
      | ^~~~~~~~~~~~~~~~~~~~~~
      | %define api.prefix {ematch_}
emp_ematch.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
    CC       emp_ematch.yacc.o
    LEX      emp_ematch.lex.c
    CC       emp_ematch.lex.o
    CC       q_fifo.o
    CC       q_sfq.o
    CC       q_red.o
    CC       q_prio.o
    CC       q_skbprio.o
    CC       q_tbf.o
    CC       q_cbq.o
    CC       q_rr.o
    CC       q_multiq.o
    CC       q_netem.o
    CC       q_choke.o
    CC       q_sfb.o
    CC       f_rsvp.o
    CC       f_u32.o
f_u32.c: In function ‘u32_parse_opt’:
f_u32.c:1113:24: warning: array subscript 0 is outside the bounds of an interior zero-length array ‘struct tc_u32_key[0]’ [-Wzero-length-bounds]
 1113 |    hash = sel2.sel.keys[0].val & sel2.sel.keys[0].mask;
      |           ~~~~~~~~~~~~~^~~
In file included from tc_util.h:10,
                 from f_u32.c:26:
../include/uapi/linux/pkt_cls.h:183:20: note: while referencing ‘keys’
  183 |  struct tc_u32_key keys[0];
      |                    ^~~~
f_u32.c:1113:47: warning: array subscript 0 is outside the bounds of an interior zero-length array ‘struct tc_u32_key[0]’ [-Wzero-length-bounds]
 1113 |    hash = sel2.sel.keys[0].val & sel2.sel.keys[0].mask;
      |                                  ~~~~~~~~~~~~~^~~
In file included from tc_util.h:10,
                 from f_u32.c:26:
../include/uapi/linux/pkt_cls.h:183:20: note: while referencing ‘keys’
  183 |  struct tc_u32_key keys[0];
      |                    ^~~~
    CC       f_route.o
    CC       f_fw.o
    CC       f_basic.o
    CC       f_bpf.o
    CC       f_flow.o
    CC       f_cgroup.o
    CC       f_flower.o
    CC       q_dsmark.o
    CC       q_gred.o
    CC       f_tcindex.o
    CC       q_ingress.o
    CC       q_hfsc.o
    CC       q_htb.o
    CC       q_drr.o
    CC       q_qfq.o
    CC       m_gact.o
    CC       m_mirred.o
    CC       m_nat.o
    CC       m_pedit.o
    CC       m_ife.o
    CC       m_skbedit.o
    CC       m_skbmod.o
    CC       m_csum.o
    CC       m_simple.o
    CC       m_vlan.o
    CC       m_connmark.o
    CC       m_bpf.o
    CC       m_tunnel_key.o
    CC       m_sample.o
    CC       p_ip.o
    CC       p_ip6.o
    CC       p_icmp.o
    CC       p_eth.o
    CC       p_tcp.o
    CC       p_udp.o
    CC       em_nbyte.o
    CC       em_cmp.o
    CC       em_u32.o
    CC       em_canid.o
    CC       em_meta.o
    CC       q_mqprio.o
    CC       q_codel.o
    CC       q_fq_codel.o
    CC       q_fq.o
    CC       q_pie.o
    CC       q_cake.o
    CC       q_hhf.o
    CC       q_clsact.o
    CC       e_bpf.o
    CC       f_matchall.o
    CC       q_cbs.o
    CC       q_etf.o
    CC       tc_core.o
    CC       tc_red.o
    CC       tc_cbq.o
    CC       tc_estimator.o
    CC       tc_stab.o
    AR       libtc.a
    LINK     tc
rm emp_ematch.lex.c

bridge
    CC       bridge.o
    CC       fdb.o
    CC       monitor.o
    CC       link.o
    CC       mdb.o
    CC       vlan.o
    LINK     bridge

misc
    CC       ss.o
    YACC     ssfilter.c
    CC       ssfilter.o
    LINK     ss
    CC       nstat
In file included from /usr/include/string.h:519,
                 from nstat.c:16:
In function ‘strncat’,
    inlined from ‘load_good_table’ at nstat.c:145:4:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:135:10: warning: ‘__builtin___strncat_chk’ output may be truncated copying 127 bytes from a string of length 4094 [-Wstringop-truncation]
  135 |   return __builtin___strncat_chk (__dest, __src, __len,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  136 |       __glibc_objsize (__dest));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~
    CC       ifstat
In file included from /usr/include/string.h:519,
                 from ifstat.c:16:
In function ‘strncpy’,
    inlined from ‘load_raw_table’ at ifstat.c:256:4:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10: warning: ‘__builtin_strncpy’ output may be truncated copying 127 bytes from a string of length 4094 [-Wstringop-truncation]
   95 |   return __builtin___strncpy_chk (__dest, __src, __len,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   96 |       __glibc_objsize (__dest));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~
    CC       rtacct
    CC       lnstat.o
    CC       lnstat_util.o
    LINK     lnstat

netem
gcc  -I../include -o maketable maketable.c -lm
gcc  -I../include -o normal normal.c -lm
gcc  -I../include -o pareto pareto.c -lm
gcc  -I../include -o paretonormal paretonormal.c -lm
./normal > normal.dist
./pareto > pareto.dist
./paretonormal > paretonormal.dist
./maketable experimental.dat > experimental.dist

genl
    CC       genl.o
    CC       ctrl.o
    LINK     genl

tipc
    CC       bearer.o
    CC       cmdl.o
    CC       link.o
    CC       media.o
    CC       misc.o
    CC       msg.o
    CC       nametable.o
    CC       node.o
    CC       socket.o
    CC       peer.o
    CC       tipc.o
    LINK     tipc

devlink
    CC       devlink.o
    CC       mnlg.o
    LINK     devlink

rdma
    CC       rdma.o
    CC       utils.o
    CC       dev.o
    CC       link.o
    CC       res.o
    LINK     rdma

man
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-address.8.in > ip-address.8
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-link.8.in > ip-link.8
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-route.8.in > ip-route.8
```
It compiles fine on `gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0` without any changes on iproute-mptcp source code.
```makefile
iproute-mptcp# ./configure && make
TC schedulers
 ATM    no

libc has setns: yes
SELinux support: no
ELF support: yes
libmnl support: no
Berkeley DB: no
need for strlcpy: yes
libcap support: no

lib
    CC       libgenl.o
    CC       libnetlink.o
libnetlink.c:120:2: warning: #warning "libmnl required for error support" [-Wcpp]
  120 | #warning "libmnl required for error support"
      |  ^~~~~~~
    AR       libnetlink.a
    CC       utils.o
    CC       rt_names.o
    CC       ll_map.o
    CC       ll_types.o
    CC       ll_proto.o
    CC       ll_addr.o
    CC       inet_proto.o
    CC       namespace.o
    CC       json_writer.o
    CC       json_print.o
    CC       names.o
    CC       color.o
    CC       bpf.o
bpf.c: In function ‘bpf_get_work_dir’:
bpf.c:779:49: warning: ‘__builtin___snprintf_chk’ output may be truncated before the last format character [-Wformat-truncation=]
  779 |  snprintf(bpf_wrk_dir, sizeof(bpf_wrk_dir), "%s/", mnt);
      |                                                 ^
In file included from /usr/include/stdio.h:867,
                 from bpf.c:14:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output between 2 and 4097 bytes into a destination of size 4096
   67 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    CC       exec.o
    CC       fs.o
    CC       dnet_ntop.o
    CC       dnet_pton.o
    CC       ipx_ntop.o
    CC       ipx_pton.o
    CC       mpls_ntop.o
    CC       mpls_pton.o
    AR       libutil.a

ip
    CC       ip.o
    CC       ipaddress.o
    CC       ipaddrlabel.o
    CC       iproute.o
    CC       iprule.o
    CC       ipnetns.o
    CC       rtm_map.o
    CC       iptunnel.o
    CC       ip6tunnel.o
    CC       tunnel.o
    CC       ipneigh.o
    CC       ipntable.o
    CC       iplink.o
    CC       ipmaddr.o
    CC       ipmonitor.o
    CC       ipmroute.o
    CC       ipprefix.o
    CC       iptuntap.o
    CC       iptoken.o
    CC       ipxfrm.o
    CC       xfrm_state.o
    CC       xfrm_policy.o
    CC       xfrm_monitor.o
    CC       iplink_dummy.o
    CC       iplink_ifb.o
    CC       iplink_nlmon.o
    CC       iplink_team.o
    CC       iplink_vcan.o
    CC       iplink_vxcan.o
    CC       iplink_vlan.o
    CC       link_veth.o
    CC       link_gre.o
    CC       iplink_can.o
    CC       iplink_xdp.o
    CC       iplink_macvlan.o
    CC       ipl2tp.o
    CC       link_vti.o
    CC       link_vti6.o
    CC       iplink_vxlan.o
    CC       tcp_metrics.o
    CC       iplink_ipoib.o
    CC       ipnetconf.o
    CC       link_ip6tnl.o
    CC       link_iptnl.o
    CC       link_gre6.o
    CC       iplink_bond.o
    CC       iplink_bond_slave.o
    CC       iplink_hsr.o
    CC       iplink_bridge.o
    CC       iplink_bridge_slave.o
    CC       ipfou.o
    CC       iplink_ipvlan.o
    CC       iplink_geneve.o
    CC       iplink_vrf.o
    CC       iproute_lwtunnel.o
    CC       ipmacsec.o
    CC       ipila.o
    CC       ipvrf.o
    CC       iplink_xstats.o
    CC       ipseg6.o
    CC       iplink_netdevsim.o
    CC       iplink_rmnet.o
    LINK     ip
    CC       rtmon.o
    LINK     rtmon

tc
    CC       tc.o
    CC       tc_qdisc.o
    CC       tc_class.o
    CC       tc_filter.o
    CC       tc_util.o
    CC       tc_monitor.o
    CC       tc_exec.o
    CC       m_police.o
    CC       m_estimator.o
    CC       m_action.o
    CC       m_ematch.o
    YACC     emp_ematch.yacc.c
emp_ematch.y:11.1-14: warning: deprecated directive: ‘%error-verbose’, use ‘%define parse.error verbose’ [-Wdeprecated]
   11 | %error-verbose
      | ^~~~~~~~~~~~~~
      | %define parse.error verbose
emp_ematch.y:12.1-22: warning: deprecated directive: ‘%name-prefix "ematch_"’, use ‘%define api.prefix {ematch_}’ [-Wdeprecated]
   12 | %name-prefix "ematch_"
      | ^~~~~~~~~~~~~~~~~~~~~~
      | %define api.prefix {ematch_}
emp_ematch.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
    CC       emp_ematch.yacc.o
    LEX      emp_ematch.lex.c
    CC       emp_ematch.lex.o
    CC       q_fifo.o
    CC       q_sfq.o
    CC       q_red.o
    CC       q_prio.o
    CC       q_skbprio.o
    CC       q_tbf.o
    CC       q_cbq.o
    CC       q_rr.o
    CC       q_multiq.o
    CC       q_netem.o
    CC       q_choke.o
    CC       q_sfb.o
    CC       f_rsvp.o
    CC       f_u32.o
    CC       f_route.o
    CC       f_fw.o
    CC       f_basic.o
    CC       f_bpf.o
    CC       f_flow.o
    CC       f_cgroup.o
    CC       f_flower.o
    CC       q_dsmark.o
    CC       q_gred.o
    CC       f_tcindex.o
    CC       q_ingress.o
    CC       q_hfsc.o
    CC       q_htb.o
    CC       q_drr.o
    CC       q_qfq.o
    CC       m_gact.o
    CC       m_mirred.o
    CC       m_nat.o
    CC       m_pedit.o
    CC       m_ife.o
    CC       m_skbedit.o
    CC       m_skbmod.o
    CC       m_csum.o
    CC       m_simple.o
    CC       m_vlan.o
    CC       m_connmark.o
    CC       m_bpf.o
    CC       m_tunnel_key.o
    CC       m_sample.o
    CC       p_ip.o
    CC       p_ip6.o
    CC       p_icmp.o
    CC       p_eth.o
    CC       p_tcp.o
    CC       p_udp.o
    CC       em_nbyte.o
    CC       em_cmp.o
    CC       em_u32.o
    CC       em_canid.o
    CC       em_meta.o
    CC       q_mqprio.o
    CC       q_codel.o
    CC       q_fq_codel.o
    CC       q_fq.o
    CC       q_pie.o
    CC       q_cake.o
    CC       q_hhf.o
    CC       q_clsact.o
    CC       e_bpf.o
    CC       f_matchall.o
    CC       q_cbs.o
    CC       q_etf.o
    CC       tc_core.o
    CC       tc_red.o
    CC       tc_cbq.o
    CC       tc_estimator.o
    CC       tc_stab.o
    AR       libtc.a
    LINK     tc
rm emp_ematch.lex.c

bridge
    CC       bridge.o
    CC       fdb.o
    CC       monitor.o
    CC       link.o
    CC       mdb.o
    CC       vlan.o
    LINK     bridge

misc
    CC       ss.o
    YACC     ssfilter.c
    CC       ssfilter.o
    LINK     ss
    CC       nstat
    CC       ifstat
    CC       rtacct
    CC       lnstat.o
    CC       lnstat_util.o
    LINK     lnstat

netem
gcc  -I../include -o maketable maketable.c -lm
gcc  -I../include -o normal normal.c -lm
gcc  -I../include -o pareto pareto.c -lm
gcc  -I../include -o paretonormal paretonormal.c -lm
./normal > normal.dist
./pareto > pareto.dist
./paretonormal > paretonormal.dist
./maketable experimental.dat > experimental.dist

genl
    CC       genl.o
    CC       ctrl.o
    LINK     genl

tipc
make[1]: Nothing to be done for 'all'.

devlink
make[1]: Nothing to be done for 'all'.

rdma
make[1]: Nothing to be done for 'all'.

man
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-address.8.in > ip-address.8
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-link.8.in > ip-link.8
sed "s|@SYSCONFDIR@|/etc/iproute2|g" ip-route.8.in > ip-route.8
```

matttbe commented:
Thank you for the output, I better understand.

I don't think what we want is to use the same variable in different places by adding this `extern` keyword. Instead, we might prefer to make them `static`.

Ideally, for maintenance purpose, we should copy what is done upstream not to diverge on non MPTCP related code. Here, I see they added `static` in different places:
- https://github.com/shemminger/iproute2/commit/fa92d8cb09e6a254d26834a66e4645ece6662f1a#diff-2860aa8df1a2e93b7f944650589f283533a0bd83d5d51f5b7cb9ba5f32ef0fe5
- https://github.com/shemminger/iproute2/commit/578cadcc68035943062d74507c3b7989741b353f
- https://github.com/shemminger/iproute2/commit/31f45088c9c891542e38144d0a1bba124934693c#diff-bb96cd5d3541f9b38e4919f956e320f99f6e66f1e205a4543f0bc8f7f0ab5f29

May you add the same `static` please?

Commit:
build: fix multiple definitions

Fix GCC release 10 being unable to compile iproute-mptcp because of multiple definition of functions.

/usr/bin/ld: ipxfrm.o:(.bss+0x0): multiple definition of `filter'; ipmroute.o:(.bss+0x0): first defined here
/usr/bin/ld: xfrm_monitor.o:(.bss+0x0): multiple definition of `listen_all_nsid'; ipmonitor.o:(.bss+0x0): first defined here
/usr/bin/ld: libutil.a(utils.o):(.bss+0xc): multiple definition of `pretty'; tipc.o:tipc.c:28: first defined here

References: https://bugzilla.opensuse.org/1160244

Signed-off-by: Arınç ÜNAL <arinc.unal@gmail.com>

arinc9 commented:
All done.

matttbe reviewed (approved):
Thank you for the update! Easier to maintain like that.
